src/ai/rival-team-gen.ts

```diff
@@ -0,0 +1,309 @@
+import type { PartyMemberStrength } from "#enums/party-member-strength";
+import { MAX_POKEMON_TYPE, PokemonType } from "#enums/pokemon-type";
+import type { SpeciesId } from "#enums/species-id";
+import { TrainerSlot } from "#enums/trainer-slot";
+import type { EnemyPokemon } from "#field/pokemon";
+import { RIVAL_6_POOL, type RivalPoolConfig } from "#trainers/rival-party-config";
+import type { Mutable } from "#types/type-helpers";
+import { applyChallenges } from "#utils/challenge-utils";
+import { NumberHolder, randSeedItem } from "#utils/common";
+import { getEnumValues } from "#utils/enums";
+import { getPokemonSpecies } from "#utils/pokemon-utils";
+
+/** The maximum number of shared weaknesses to tolerate when balancing weakness */
+const MAX_SHARED_WEAKNESSES = 2;
+const MAX_SHARED_TYPES = 1;
+
+/** Record of the chosen indices in the rival species pool, for type balancing based on the final fight */
+const CHOSEN_RIVAL_ROLLS: (undefined | [number] | [number, number])[] = new Array(6);
+
+/**
+ * Return the species from the rival species pool based on a previously chosen roll
+ * @param param0 - The chosen rolls for the rival species pool
+ * @param pool - The rival species pool for the slot
+ * @returns - A `SpeciesId` if found, or `undefined` if no species exists at the chosen roll
+ */
+function rivalRollToSpecies(
+  [roll1, roll2]: [number] | [number, number],
+  pool: readonly (SpeciesId | readonly SpeciesId[])[],
+): SpeciesId | undefined {
+  // Drop readonly because typescript does not preserve it with its `isArray` signature
+  const pull1 = pool[roll1] as Mutable<(typeof pool)[number]>;
+  if (!Array.isArray(pull1)) {
+    return pull1;
+  }
+  if (roll2 == null) {
+    return;
+  }
+
+  return pull1[roll2];
+}
+
+/**
+ * Calculates the types that the given species is weak to
+ *
+ * @remarks
+ * - Considers Levitate as a guaranteed ability if the species has it as all 3 possible abilities.
+ * - Accounts for type effectiveness challenges via {@linkcode applyChallenges}
+ *
+ * @privateRemarks
+ * Despite potentially being a useful utility method, this is intentionally *not*
+ * exported because it uses logic specific to this file, such as excluding the second type for Tera starters
+ *
+ * @param species - The species to calculate weaknesses for
+ * @param exclude2ndType - (default `false`) Whether to exclude the second type when calculating weaknesses
+ *  Intended to be used for starters since they will terastallize to their primary type.
+ * @returns The set of types that the species is weak to
+ */
+function getWeakTypes(species: PokemonSpecies, exclude2ndType = false): Set<PokemonType> {
+  const weaknesses = new Set<PokemonType>();
+  const alwaysHasLevitate =
+    species.ability1 === AbilityId.LEVITATE
+    && (species.ability2 == null || species.ability2 !== AbilityId.LEVITATE)
+    && (species.abilityHidden == null || species.abilityHidden !== AbilityId.LEVITATE);
```

Comment:
```suggestion
  const alwaysHasLevitate =
    species.ability1 === AbilityId.LEVITATE
    && (species.ability2 == null || species.ability2 === AbilityId.LEVITATE)
    && (species.abilityHidden == null || species.abilityHidden === AbilityId.LEVITATE);
```
It should be checking that all existing abilities are Levitate right?